core/dbt/include/global_project/macros/materializations/common/merge.sql

```diff
@@ -20,16 +20,33 @@
 {% endmacro %}
 
 
-{% macro common_get_merge_sql(target, source, unique_key, dest_columns) -%}
+{% macro common_get_merge_sql(target, source, unique_key, dest_columns, dest_partition) -%}
     {%- set dest_cols_csv =  get_quoted_csv(dest_columns | map(attribute="name")) -%}
 
     merge into {{ target }} as DBT_INTERNAL_DEST
     using {{ source }} as DBT_INTERNAL_SOURCE
 
+    {% set conditions = [] %}
+    
+    {% if dest_partition %}
+        {% set dest_partition_filter %}
+            DBT_INTERNAL_DEST.{{dest_partition.name}}
+                between '{{dest_partition.min}}' and '{{dest_partition.max}}'
+        {% endset %}
+        {% do conditions.append(dest_partition_filter) %}
+    {% endif %}
+
     {% if unique_key %}
-        on DBT_INTERNAL_SOURCE.{{ unique_key }} = DBT_INTERNAL_DEST.{{ unique_key }}
-    {% else %}
-        on FALSE
+        {% set unique_key_match %}
+            DBT_INTERNAL_SOURCE.{{ unique_key }} = DBT_INTERNAL_DEST.{{ unique_key }}
+        {% endset %}
+        {% do conditions.append(unique_key_match) %}
+    {% endif %}
+    
+        ON {{conditions|join(' and ')}}
+    
+    {% if conditions|length == 0 %}
```

Comment:
This logic makes sense, but it's a bit hard to follow. Is there a more explicit way for us to implement this logic? Maybe something that looks more like:

```
merge into {{ target }} as DBT_INTERNAL_DEST
    using {{ source }} as DBT_INTERNAL_SOURCE
on 
{% if conditions | length == 0 %}
  FALSE 
{% else %}
  {{ conditions | join(' and ') }}
{% endif %}
```